### PR TITLE
feat: storing and exposing bpmXml on Tasklist V1 

### PR DESCRIPTION
**Pull request by marcosgvieira**
## Description

- Adding a new field on Process Index to store the bpmn xml;
- Exposing the new field into a new API method v1/internal/processes/{processDefinitionKey};
- New field is being ignored in v1/internal/processes/search to avoid a big response and bpmnXml should not be used there.

## Related issues

Closes https://github.com/camunda/tasklist/issues/4900

### CODE REVIEW

**Comment by marcosgvieira** [comment]
@nathansandi this was created from cherry-picks from https://github.com/camunda/zeebe/pull/17607

File: tasklist/webapp/src/main/java/io/camunda/tasklist/webapp/security/identity/IdentityAuthorizationService.java [hunk 1]
```diff
@@ -102,4 +71,76 @@ public boolean isAllowedToStartProcess(final String processDefinitionKey) {
...
+
+  public List<String> getProcessReadFromAuthorization() {
+    return getFromAuthorization(IdentityAuthorization.PROCESS_PERMISSION_READ);
+  }
+
+  public List<String> getProcessDefinitionsFromAuthorization() {
+    return getFromAuthorization(IdentityAuthorization.PROCESS_PERMISSION_START);
+  }
+
+  private List<String> getFromAuthorization(final String type) {
```
Comment (douglasbouttell-camunda) [inline tasklist/webapp/src/main/java/io/camunda/tasklist/webapp/security/identity/IdentityAuthorizationService.java:83]: **Restructuring suggestion:** It looks like the common pattern here is get an IdentityAuthorization object and then select the correct list of permissions. Could we simplify this to 

```java
@Nullable
private IdentityAuthorization getIdentityAuthorization() {
  if (!tasklistPropertires.getIdentity().isResourcePermissionsEnabled()) {
    return null;
  }
  if (tasklistProperties.getIdentity().getBaseUrl() == null) {
    return null;
  }

  final Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
  return switch (authentication) {
    case IdentityAuthentication identityAuthentication -> identityAuthentication.getAuthorizations()
    case JwtAuthenticationToken jwtAuthenticationToken -> {
      final Identity = SpringContextHolder.getBean(Identity.class);
      yield new IdentityAuthorization(identity.authorizations().forToken(jwtAuthenticationToken.getToken().getTokenValue())
    }
    case TokenAuthentication tokenAuthentication -> {
      final Identity = SpringContextHolder.getBean(Identity.class);
      yield new IdentityAuthorization(
        identity
          .authorizations()
          .forToken(tokenAuthentication.getAccessToken(), tokenAuthentication.getOrganization()))
    }
    default -> null
  }
}

private List<String> getFromAuthorization(final String type) {
  final IdentityAuthorization auth = getIdentityAuthorization();
  if (auth == null) {
    return singletonList(IdentityProperties.ALL_RESOURCES);
  }
  return switch (type) {
    case IdentityAuthorization.PROCESS_PERMISSION_READ -> auth.getProcessAllowedToRead()
    case IdentityAuthorization.PROCESS_PERMISSION_START -> auth.getProcessAllowedToStart()
    default -> emptyList()
  }
}
```

File: tasklist/webapp/src/main/java/io/camunda/tasklist/webapp/security/identity/IdentityAuthorizationService.java [hunk 2]
```diff
@@ -102,4 +71,76 @@ public boolean isAllowedToStartProcess(final String processDefinitionKey) {
...
+                              ((TokenAuthentication) authentication).getAccessToken(),
+                              ((TokenAuthentication) authentication).getOrganization()))
+                  .getProcessesAllowedToStart();
+          default -> null;
+        };
+      }
+    }
+    final List<String> result = new ArrayList<String>();
+    result.add(IdentityProperties.ALL_RESOURCES);
+    return result;
```
[the remark this answers is not in the document]
Comment (douglasbouttell-camunda) [reply tasklist/webapp/src/main/java/io/camunda/tasklist/webapp/security/identity/IdentityAuthorizationService.java:144]: Could even do `List.of`

File: tasklist/webapp/src/main/java/io/camunda/tasklist/webapp/api/rest/v1/controllers/internal/ProcessInternalController.java [hunk 3]
```diff
@@ -73,6 +73,27 @@ public class ProcessInternalController extends ApiErrorController {
...
 
+  @Operation(
+      summary = "Returns the process by ProcessDefinitionKey",
+      description = "Returns the process by ProcessDefinitionKey",
+      responses = {
+        @ApiResponse(
+            description = "On success returned",
+            responseCode = "200",
+            useReturnTypeSchema = true),
+        @ApiResponse(description = "Process Not Found", responseCode = "404")
```
[the remark this answers is not in the document]
Comment (marcosgvieira) [reply tasklist/webapp/src/main/java/io/camunda/tasklist/webapp/api/rest/v1/controllers/internal/ProcessInternalController.java:78]: Thanks @douglasbouttell-camunda I added it!

File: tasklist/webapp/src/main/java/io/camunda/tasklist/webapp/api/rest/v1/entities/ProcessResponse.java [hunk 4]
```diff
@@ -109,8 +112,34 @@ public ProcessResponse setTenantId(String tenantId) {
...
+        .setVersion(process.getVersion())
+        .setStartEventFormId(startEventFormId)
+        .setTenantId(process.getTenantId())
+        .setBpmnXml(process.getBpmnXml());
+  }
+
+  public static ProcessResponse fromProcessEntityWithoutBpmnXml(
+      ProcessEntity process, String startEventFormId) {
+    process.setBpmnXml(null);
+    return fromProcessEntity(process, startEventFormId);
```
[the remark this answers is not in the document]
Comment (marcosgvieira) [reply tasklist/webapp/src/main/java/io/camunda/tasklist/webapp/api/rest/v1/entities/ProcessResponse.java]: I changed the implementation a bit, I prefer to have by default everything being used, check and let me know wdyt.

File: tasklist/webapp/src/main/java/io/camunda/tasklist/webapp/security/identity/IdentityAuthorization.java [hunk 5]
```diff
@@ -24,22 +24,34 @@
...
-          break;
-        } else {
-          processesAllowedToStart.add(authorization.getResourceKey());
+    processesAllowedToRead = new ArrayList<String>();
+    for (final Authorization authorization : authorizations) {
+      if (authorization.getResourceType().equals(PROCESS_DEFINITION)) {
+        if (authorization.getPermissions().contains(PROCESS_PERMISSION_START)) {
+          if (authorization.getResourceKey().equals(IdentityProperties.ALL_RESOURCES)) {
+            processesAllowedToStart.clear();
+            processesAllowedToStart.add(IdentityProperties.ALL_RESOURCES);
```
[the remark this answers is not in the document]
Comment (marcosgvieira) [reply tasklist/webapp/src/main/java/io/camunda/tasklist/webapp/security/identity/IdentityAuthorization.java]: It's not possible to break as we can miss another types of authorizations, I did refactor the code to keep everything in the same loop (I have a concert that in the future we will have a lot of access restrictions and I don't want to loop on this list).

File: tasklist/webapp/src/main/java/io/camunda/tasklist/webapp/security/identity/IdentityAuthorizationService.java [hunk 6]
```diff
@@ -102,4 +71,76 @@ public boolean isAllowedToStartProcess(final String processDefinitionKey) {
...
+          case IdentityAuthorization.PROCESS_PERMISSION_START ->
+              new IdentityAuthorization(
+                      identity
+                          .authorizations()
+                          .forToken(jwtAuthenticationToken.getToken().getTokenValue()))
+                  .getProcessesAllowedToStart();
+          default -> null;
+        };
+
+      } else if (authentication instanceof TokenAuthentication) {
```
[the remark this answers is not in the document]
Comment (marcosgvieira) [reply tasklist/webapp/src/main/java/io/camunda/tasklist/webapp/security/identity/IdentityAuthorizationService.java]: @douglasbouttell-camunda super nice, I added this.

File: tasklist/webapp/src/main/java/io/camunda/tasklist/webapp/security/identity/IdentityAuthorizationService.java [hunk 7]
```diff
@@ -102,4 +71,76 @@ public boolean isAllowedToStartProcess(final String processDefinitionKey) {
...
+        return switch (type) {
+          case IdentityAuthorization.PROCESS_PERMISSION_READ ->
+              identityAuthentication.getAuthorizations().getProcessesAllowedToRead();
+          case IdentityAuthorization.PROCESS_PERMISSION_START ->
+              identityAuthentication.getAuthorizations().getProcessesAllowedToStart();
+          default -> null;
+        };
+      } else if (authentication instanceof JwtAuthenticationToken) {
+        final JwtAuthenticationToken jwtAuthenticationToken =
+            (JwtAuthenticationToken) authentication;
```
[the remark this answers is not in the document]
Comment (marcosgvieira) [reply tasklist/webapp/src/main/java/io/camunda/tasklist/webapp/security/identity/IdentityAuthorizationService.java]: @douglasbouttell-camunda super nice, I added this.

File: tasklist/webapp/src/main/java/io/camunda/tasklist/webapp/security/identity/IdentityAuthorizationService.java [hunk 8]
```diff
@@ -102,4 +71,76 @@ public boolean isAllowedToStartProcess(final String processDefinitionKey) {
...
+    return getFromAuthorization(IdentityAuthorization.PROCESS_PERMISSION_START);
+  }
+
+  private List<String> getFromAuthorization(final String type) {
+    if (tasklistProperties.getIdentity().isResourcePermissionsEnabled()
+        && tasklistProperties.getIdentity().getBaseUrl() != null) {
+      final Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
+      if (authentication instanceof IdentityAuthentication) {
+        final IdentityAuthentication identityAuthentication =
+            (IdentityAuthentication) authentication;
```
[the remark this answers is not in the document]
Comment (marcosgvieira) [reply tasklist/webapp/src/main/java/io/camunda/tasklist/webapp/security/identity/IdentityAuthorizationService.java]: @douglasbouttell-camunda super nice, I added this.

File: tasklist/webapp/src/main/java/io/camunda/tasklist/webapp/security/identity/IdentityAuthorizationService.java (hunk 1 above)
Comment (marcosgvieira) [reply tasklist/webapp/src/main/java/io/camunda/tasklist/webapp/security/identity/IdentityAuthorizationService.java]: @douglasbouttell-camunda I liked it, refactor implemented with minor changes on your suggestion.

File: tasklist/importer-850/src/main/java/io/camunda/tasklist/zeebeimport/v850/processors/es/ProcessZeebeRecordProcessorElasticSearch.java [hunk 9]
```diff
@@ -115,6 +115,7 @@ private ProcessEntity createEntity(
     processEntity.setBpmnProcessId(process.getBpmnProcessId());
     processEntity.setVersion(process.getVersion());
     processEntity.setTenantId(process.getTenantId());
+    processEntity.setBpmnXml(new String(process.getResource()));
```
Comment (nathansandi) [inline tasklist/importer-850/src/main/java/io/camunda/tasklist/zeebeimport/v850/processors/es/ProcessZeebeRecordProcessorElasticSearch.java:118]: Should we add the Encoding here?

`+processEntity.setBpmnXml(new String(process.getResource(), StandardCharsets.UTF_8));`

File: tasklist/webapp/src/main/java/io/camunda/tasklist/webapp/api/rest/v1/entities/ProcessResponse.java [hunk 10]
```diff
@@ -112,51 +158,10 @@ public boolean equals(Object o) {
...
-
-  @Override
-  public int hashCode() {
-    int result = Objects.hash(id, name, bpmnProcessId, version, startEventFormId, tenantId);
-    result = 31 * result + Arrays.hashCode(sortValues);
-    return result;
-  }
-
-  @Override
-  public String toString() {
```
Comment (nathansandi) [inline tasklist/webapp/src/main/java/io/camunda/tasklist/webapp/api/rest/v1/entities/ProcessResponse.java:129]: [Question] Is there a reason for delete the toString method?

File: tasklist/importer-850/src/main/java/io/camunda/tasklist/zeebeimport/v850/processors/es/ProcessZeebeRecordProcessorElasticSearch.java (hunk 9 above)
Comment (marcosgvieira) [reply tasklist/importer-850/src/main/java/io/camunda/tasklist/zeebeimport/v850/processors/es/ProcessZeebeRecordProcessorElasticSearch.java:118]: I don't see the need, UTF-8 is set the default, do you think we should explicitly do this?

File: tasklist/webapp/src/main/java/io/camunda/tasklist/webapp/api/rest/v1/entities/ProcessResponse.java (hunk 10 above)
Comment (marcosgvieira) [reply tasklist/webapp/src/main/java/io/camunda/tasklist/webapp/api/rest/v1/entities/ProcessResponse.java]: @nathansandi I removed by mistake, I put it back :)

File: tasklist/importer-850/src/main/java/io/camunda/tasklist/zeebeimport/v850/processors/es/ProcessZeebeRecordProcessorElasticSearch.java (hunk 9 above)
Comment (nathansandi) [reply tasklist/importer-850/src/main/java/io/camunda/tasklist/zeebeimport/v850/processors/es/ProcessZeebeRecordProcessorElasticSearch.java:118]: If it is default, not necessary (Didnt know this information 💯 ) 

Thank you for clarifying this.

**Review by nathansandi — APPROVED** [review]
From backend side, looks good on my side 
💯

**Review by douglasbouttell-camunda — APPROVED** [review]
LGTM. Just need to fix your commit messages

**Pull request merged**